Service registration fails after WebSocket reconnection: Service id must contain ':'

## Bug: Service registration fails after WebSocket reconnection

### Symptoms
After a daemon has been running for some time and experiences a WebSocket reconnection, new `registerService()` calls fail with:

```
AssertionError: Service id info must contain ':'
```

Error trace (server-side, Python):
```python
File "/home/hypha/core/workspace.py", line 2071, in register_service
    assert ":" in service.id, "Service id info must contain ':'"
```

### Expected Behavior
`registerService()` should work reliably after reconnection, constructing the full service ID as `{workspace}/{client_id}:{service_id}`.

### Actual Behavior
- **Fresh connection**: Works perfectly. The ID is correctly constructed as `ws-user-github|478667/clientId:svamp-session-xxx` (contains both `/` and `:`).
- **After reconnection**: The server receives a service ID **without** the `:` separator. The server prepends the workspace (`{ws}/{service_id}`), resulting in something like `ws-user-github|478667/svamp-session-xxx` — which fails the assertion.

### Analysis

The JS client's `_extract_service_info()` (rpc.js line 1652) constructs:
```javascript
id: `${config.workspace}/${this._client_id}:${service["id"]}`
```

This should always include `:`. However, after reconnection, the server receives an ID without `:`, suggesting either:
1. `_client_id` is somehow lost/undefined after reconnection
2. The `_extract_service_info` method is not being called for newly registered services
3. The service info is being serialized differently after reconnection
4. The server's workspace manager proxy (`manager.registerService`) is routing to a stale target

### Reproduction Context
- **hypha-rpc**: 0.21.27 (JS) + 0.21.27 (Python)
- **hypha server**: 0.21.62
- **Scenario**: Long-running daemon process (svamp-cli) connected via WebSocket. Machine service and debug service remain working after reconnection. Only NEW service registrations fail.
- **Workaround**: Restarting the daemon (fresh connection) temporarily fixes the issue.

### Steps to Reproduce
1. Connect to Hypha server via WebSocket
2. Register a service (works fine)
3. Wait for WebSocket disconnection + automatic reconnection
4. Try to register a new service → fails with the assertion error

### Key Observation
Services registered BEFORE reconnection continue to work (they were re-registered during the reconnection `services_registered` handler). Only services registered AFTER reconnection (via new `registerService()` calls from application code) fail.

This suggests the issue is in the `register_service()` code path (rpc.js line 1668) rather than the reconnection re-registration path (rpc.js line 601).

--- a/javascript/src/http-client.js
+++ b/javascript/src/http-client.js
@@ -730,25 +730,25 @@ export async function _connectToServerHTTP(config) {
   // Auto-refresh workspace manager proxy after reconnection.
   // See websocket-client.js for detailed explanation.
   let isInitialRefresh = true;
-  rpc.on("manager_refreshed", async ({ manager: internalManager }) => {
+  rpc.on("manager_refreshed", async () => {
     if (isInitialRefresh) {
       isInitialRefresh = false;
       return;
     }
     try {
-      const freshWm = internalManager;
-      for (const key of Object.keys(freshWm)) {
-        if (typeof freshWm[key] === "function") {
-          wm[key] = freshWm[key];
+      const newTarget = `*/${rpc._connection.manager_id}`;
+      for (const key of Object.keys(wm)) {
+        if (typeof wm[key] === "function" && wm[key].__rpc_object__) {
+          wm[key].__rpc_object__._rtarget = newTarget;
         }
       }
       console.info(
-        "Workspace manager proxy refreshed after reconnection (new manager_id:",
+        "Workspace manager proxy retargeted after reconnection (new manager_id:",
         rpc._connection?.manager_id + ")",
       );
     } catch (err) {
       console.warn(
-        "Failed to refresh workspace manager after reconnection:",
+        "Failed to retarget workspace manager after reconnection:",
         err,
       );
     }

--- a/javascript/src/rpc.js
+++ b/javascript/src/rpc.js
@@ -2325,6 +2325,10 @@ export class RPC extends MessageEmitter {
     function remote_method() {
       return new Promise(async (resolve, reject) => {
         try {
+          // Read target_id from encoded_method at call time (not captured
+          // at generation time) so that _rtarget can be updated after
+          // reconnection without regenerating the method closures.
+          const target_id = encoded_method._rtarget;
           let local_session_id = randId();
           if (local_parent) {
             // Store the children session under the parent

--- a/javascript/src/websocket-client.js
+++ b/javascript/src/websocket-client.js
@@ -838,50 +838,34 @@ export async function connectToServer(config) {
   wm.rpc = rpc;
 
   // Auto-refresh workspace manager proxy after reconnection.
-  // When the server restarts, it assigns a new manager_id. The wm proxy
-  // returned to the caller has methods bound to the old manager_id.
-  //
-  // The RPC layer fires "manager_refreshed" IMMEDIATELY after getting the
-  // fresh manager service — before service re-registration. This minimizes
-  // the window where stale methods exist (~100ms instead of ~2-3s).
-  //
-  // Combined with the RPC layer's immediate rejection of pending calls to
-  // the old manager_id, recovery is near-instant.
+  // When the server restarts, it assigns a new manager_id. The remote
+  // methods on wm have their target baked into __rpc_object__._rtarget.
+  // Since remote_method() reads _rtarget at call time (not generation
+  // time), we just update _rtarget on every existing method — no need
+  // to regenerate methods or copy from a fresh proxy. This preserves
+  // locally-overridden methods (registerService, unregisterService, etc.)
+  // that would otherwise be lost by wholesale function copying.
   let isInitialRefresh = true;
-  rpc.on("manager_refreshed", async ({ manager: internalManager }) => {
+  rpc.on("manager_refreshed", async () => {
     if (isInitialRefresh) {
       isInitialRefresh = false;
       return; // Skip the first event (initial connection, wm is already fresh)
     }
     try {
-      let freshWm;
-      if (config.kwargs_expansion) {
-        // kwargs_expansion changes the method signatures, so we need to
-        // fetch a new manager with matching config
-        freshWm = await rpc.get_manager_service({
-          timeout: config.method_timeout || 30,
-          case_conversion: "camel",
-          kwargs_expansion: config.kwargs_expansion,
-        });
-      } else {
-        // The internal manager already uses case_conversion: "camel",
-        // so we can copy directly without an extra RPC call
-        freshWm = internalManager;
-      }
-      // Copy all function properties from fresh wm onto existing wm object.
-      // This preserves the caller's reference while updating method targets.
-      for (const key of Object.keys(freshWm)) {
-        if (typeof freshWm[key] === "function") {
-          wm[key] = freshWm[key];
+      const newTarget = `*/${rpc._connection.manager_id}`;
+      // Retarget all remote methods on the wm proxy to the new manager
+      for (const key of Object.keys(wm)) {
+        if (typeof wm[key] === "function" && wm[key].__rpc_object__) {
+          wm[key].__rpc_object__._rtarget = newTarget;
         }
       }
       console.info(
-        "Workspace manager proxy refreshed after reconnection (new manager_id:",
+        "Workspace manager proxy retargeted after reconnection (new manager_id:",
         rpc._connection?.manager_id + ")",
       );
     } catch (err) {
       console.warn(
-        "Failed to refresh workspace manager after reconnection:",
+        "Failed to retarget workspace manager after reconnection:",
         err,
       );
     }

--- a/python/hypha_rpc/http_client.py
+++ b/python/hypha_rpc/http_client.py
@@ -664,26 +664,23 @@ async def _connect_to_server_http(config: dict):
             _is_initial_refresh["value"] = False
             return
         try:
-            internal_manager = event_data.get("manager") if isinstance(event_data, dict) else None
-            if internal_manager:
-                fresh_wm = internal_manager
-            else:
-                fresh_wm = await rpc.get_manager_service(
-                    {"timeout": config.get("method_timeout", 30), "case_conversion": "snake"}
-                )
-            for key in dir(fresh_wm):
+            new_target = f"*/{connection.manager_id}"
+            # Retarget remote methods to new manager (see websocket_client.py)
+            for key in dir(wm):
                 if key.startswith("_"):
                     continue
-                val = getattr(fresh_wm, key, None)
-                if callable(val):
-                    setattr(wm, key, val)
+                val = getattr(wm, key, None)
+                if callable(val) and hasattr(val, "_encoded_method"):
+                    val._encoded_method["_rtarget"] = new_target
+                    if hasattr(val, "__rpc_object__"):
+                        val.__rpc_object__["_rtarget"] = new_target
             logger.info(
-                "Workspace manager proxy refreshed after reconnection (new manager_id: %s)",
+                "Workspace manager proxy retargeted after reconnection (new manager_id: %s)",
                 getattr(connection, "manager_id", "unknown"),
             )
         except Exception as err:
             logger.warning(
-                "Failed to refresh workspace manager after reconnection: %s", err
+                "Failed to retarget workspace manager after reconnection: %s", err
             )
 
     rpc.on("manager_refreshed", _refresh_wm_proxy)

--- a/python/hypha_rpc/websocket_client.py
+++ b/python/hypha_rpc/websocket_client.py
@@ -971,30 +971,27 @@ async def _connect_to_server(config):
             _is_initial_refresh["value"] = False
             return  # Skip the first event (initial connection, wm is already fresh)
         try:
-            # The internal manager already uses case_conversion: "snake",
-            # so we can copy directly without an extra RPC call
-            internal_manager = event_data.get("manager") if isinstance(event_data, dict) else None
-            if internal_manager:
-                fresh_wm = internal_manager
-            else:
-                fresh_wm = await rpc.get_manager_service(
-                    {"timeout": config.get("method_timeout", 30), "case_conversion": "snake"}
-                )
-            # Copy all callable attributes from fresh wm onto existing wm object.
-            # This preserves the caller's reference while updating method targets.
-            for key in dir(fresh_wm):
+            new_target = f"*/{connection.manager_id}"
+            # Retarget all remote methods on the wm proxy to the new manager.
+            # Remote methods read _encoded_method["_rtarget"] at call time,
+            # so updating _rtarget is sufficient — no need to regenerate
+            # methods or copy from a fresh proxy. This preserves locally-
+            # overridden methods (register_service, unregister_service, etc.).
+            for key in dir(wm):
                 if key.startswith("_"):
                     continue
-                val = getattr(fresh_wm, key, None)
-                if callable(val):
-                    setattr(wm, key, val)
+                val = getattr(wm, key, None)
+                if callable(val) and hasattr(val, "_encoded_method"):
+                    val._encoded_method["_rtarget"] = new_target
+                    if hasattr(val, "__rpc_object__"):
+                        val.__rpc_object__["_rtarget"] = new_target
             logger.info(
-                "Workspace manager proxy refreshed after reconnection (new manager_id: %s)",
+                "Workspace manager proxy retargeted after reconnection (new manager_id: %s)",
                 getattr(connection, "manager_id", "unknown"),
             )
         except Exception as err:
             logger.warning(
-                "Failed to refresh workspace manager after reconnection: %s", err
+                "Failed to retarget workspace manager after reconnection: %s", err
             )
 
     rpc.on("manager_refreshed", _refresh_wm_proxy)

